theseus/embodied/robot/joint.py

```diff
@@ -217,3 +217,158 @@ def _rotation_impl(self, angle: torch.Tensor) -> torch.Tensor:
         rot[:, 0, 1] = -rot[:, 1, 0]
         rot[:, 1, 1] = rot[:, 0, 0]
         return rot
+
+
+class _PrismaticJointImpl(Joint):
+    def __init__(
+        self,
+        name: str,
+        parent: int = -1,
+        child: int = -1,
+        origin: Optional[torch.Tensor] = None,
+        dtype: Optional[torch.dtype] = None,
+    ):
+        super().__init__(name, parent, child, origin, dtype)
+
+    def dof(self) -> int:
+        return 1
+
+    @abc.abstractmethod
+    def _translation_impl(self, angle: torch.Tensor) -> torch.Tensor:
+        pass
+
+    def relative_pose(self, angle: torch.Tensor) -> torch.Tensor:
+        trans = self._translation_impl(angle)
+        ret = angle.new_empty(angle.shape[0], 3, 4)
+        ret[:, :, :3] = self.origin[:, :, :3]
+        ret[:, :, 3:] = (
+            self.origin[:, :, :3] @ trans.view(-1, 3, 1) + self.origin[:, :, 3:]
+        )
+        return ret
+
+
+class PrismaticJoint(_PrismaticJointImpl):
+    def __init__(
+        self,
+        name: str,
+        prismatic_axis: torch.Tensor,
+        parent: int = -1,
+        child: int = -1,
+        origin: Optional[torch.Tensor] = None,
+        dtype: Optional[torch.dtype] = None,
+    ):
+        if prismatic_axis.ndim == 1:
+            prismatic_axis = prismatic_axis.view(-1, 1)
+
+        if prismatic_axis.ndim != 2 or prismatic_axis.shape != (3, 1):
+            raise ValueError("The prismatic axis must be a 3-D vector.")
+
+        super().__init__(name, parent, child, origin, dtype)
+
+        if prismatic_axis.dtype != self.dtype:
+            raise ValueError(f"The dtype of prismatic_axis should be {self.dtype}.")
+
+        self._axis[:3] = prismatic_axis
+
+    def _translation_impl(self, angle: torch.Tensor) -> torch.Tensor:
+        if angle.ndim == 1:
+            angle = angle.view(-1, 1)
+        if angle.ndim != 2 or angle.shape[1] != 1:
+            raise ValueError("The joint angle must be a vector.")
+        return angle @ self.axis[:3].view(1, -1)
+
+
+class PrismaticJointX(_PrismaticJointImpl):
+    def __init__(
+        self,
+        name: str,
+        parent: int = -1,
+        child: int = -1,
+        origin: Optional[torch.Tensor] = None,
+        dtype: Optional[torch.dtype] = None,
+    ):
+        super().__init__(name, parent, child, origin, dtype)
+        self._axis[3] = 1
+
+    def _translation_impl(self, angle: torch.Tensor) -> torch.Tensor:
```

Comment:
The code in `_translation_impl` and `relative_pose` of these 3 classes is exactly the same except for one number, so you can just have a base class `_PrismaticJointAxis` that defines a member variable `self._axis_idx: Optional[int] = None`, which gets overridden by all base classes. Then we have 
```python
def _translation_impl(self, angle: torch.Tensor) -> torch.Tensor:
    self._check_angle_shape(angle)
    trans = angle.new_zeros(angle.shape[0], 3)
    trans[:, self._axis_idx] = angle
    return trans

def relative_pose(self, angle: torch.Tensor) -> torch.Tensor:
    self._check_angle_shape(angle)
    ret = angle.new_empty(angle.shape[0], 3, 4)
    ret[:, :, :3] = self.origin[:, :, :3]
    ret[:, :, 3] = angle * self.origin[:, :, self._axis_idx] + self.origin[:, :, 3]
    return ret
```

Reply:
Refactored. Thanks!